File: DraftRightMobile/android/app/src/main/kotlin/com/draftright/keyboard/composer/BufferingComposer.kt
package com.draftright.keyboard.composer

import com.draftright.keyboard.ComposeResult
import com.draftright.keyboard.Composer

/**
 * Shared skeleton for composers that accumulate raw input into a buffer and
 * expose a transformed "display" (composing) string — e.g. rōmaji→kana for
 * Japanese, pinyin for Chinese. Subclasses implement only [transform]; the
 * keystroke/backspace/commit/reset machinery lives here once (Rule #1).
 *
 * Not used by TelexComposer, whose Vietnamese combining happens per-keystroke
 * on the live buffer rather than as a whole-buffer transform.
 */
abstract class BufferingComposer : Composer {

    private val raw = StringBuilder()
    private var cachedRaw: String? = null
    private var cachedDisplay: String = ""

    /** Convert the accumulated raw input to the composing/display text. */
    protected abstract fun transform(raw: String): String

    /** Which characters are buffered as input. Default: letters. */
    protected open fun isInputChar(char: Char): Boolean = char.isLetter()

    /** Max raw length before auto-committing, to bound the buffer. */
    protected open val maxLen: Int = 48

    /** Memoized transform — recomputed only when the raw buffer changes, so
     *  currentComposingText()/onKey don't re-run the conversion every call. */
    private fun display(): String {
        val r = raw.toString()
        if (cachedRaw != r) {
            cachedRaw = r
            cachedDisplay = transform(r)
        }
        return cachedDisplay
    }

    override fun onKey(char: Char): ComposeResult {
        if (raw.length >= maxLen) {
            val committed = display()
            raw.clear()
            if (isInputChar(char)) {
                raw.append(char)
                return ComposeResult.Composing(display())
            }
            return ComposeResult.Commit(committed + char)
        }
        if (!isInputChar(char)) {
            if (raw.isEmpty()) return ComposeResult.PassThrough
            val out = display() + char
            raw.clear()
            return ComposeResult.Commit(out)
        }
        raw.append(char)
        return ComposeResult.Composing(display())
    }

    override fun onBackspace(): ComposeResult {
        if (raw.isEmpty()) return ComposeResult.PassThrough
        raw.deleteCharAt(raw.length - 1)
        return if (raw.isEmpty()) ComposeResult.Consumed else ComposeResult.Composing(display())
    }

    override fun reset() {
        raw.clear()
        cachedRaw = null
        cachedDisplay = ""
    }

    override fun currentComposingText(): String = display()
}

File: DraftRightMobile/ios/DraftRightKeyboardCore/Sources/DraftRightKeyboardCore/Composer/BufferingComposer.swift
import Foundation

/// Shared skeleton for composers that accumulate raw input into a buffer and
/// expose a transformed "display" (composing) string — e.g. rōmaji→kana for
/// Japanese, pinyin for Chinese. Subclasses override only `transform`; the
/// keystroke/backspace/commit/reset machinery lives here once (Rule #1).
///
/// Mirrors the Kotlin BufferingComposer. Not used by TelexComposer (its
/// Vietnamese combining is per-keystroke, not a whole-buffer transform).
open class BufferingComposer: Composer {

    private var raw = ""
    private var cachedRaw: String?
    private var cachedDisplay = ""

    public init() {}

    /// Convert the accumulated raw input to the composing/display text.
    /// Subclasses MUST override.
    open func transform(_ raw: String) -> String { raw }

    /// Which characters are buffered as input. Default: letters.
    open func isInputChar(_ char: Character) -> Bool { char.isLetter }

    /// Max raw length before auto-committing, to bound the buffer.
    open var maxLen: Int { 48 }

    /// Memoized transform — recomputed only when the raw buffer changes.
    private func display() -> String {
        if cachedRaw != raw {
            cachedRaw = raw
            cachedDisplay = transform(raw)
        }
        return cachedDisplay
    }

    public func onKey(_ char: Character) -> ComposeResult {
        if raw.count >= maxLen {
            let committed = display()
            raw = ""
            if isInputChar(char) {
                raw.append(char)
                return .composing(display())
            }
            return .commit(committed + String(char))
        }
        if !isInputChar(char) {
            if raw.isEmpty { return .passThrough }
            let out = display() + String(char)
            raw = ""
            return .commit(out)
        }
        raw.append(char)
        return .composing(display())
    }

    public func onBackspace() -> ComposeResult {
        if raw.isEmpty { return .passThrough }
        raw.removeLast()
        return raw.isEmpty ? .consumed : .composing(display())
    }

    public func reset() {
        raw = ""
        cachedRaw = nil
        cachedDisplay = ""
    }

    public func currentComposingText() -> String { display() }
}
